path/path_test.go

```diff
@@ -28,3 +28,27 @@ func TestPathParsing(t *testing.T) {
 		}
 	}
 }
+
+func TestPathCleaning(t *testing.T) {
+	cases := map[string]string{
+		// .. should strip the preceding path segment
```

Comment:
I can see similar code being useful in go-ipld, but I think `github.com/ipfs/go-ipfs/path` needs to resolve `.` and `..` references on parse.

This is required for URIs – [even absolute URIs](http://tools.ietf.org/html/rfc3986#section-6.2.2.3) – so web servers including Go's `net/http` [perform dot-related normalization automatically](https://github.com/golang/go/blob/6fd82d83d697edebbe826a028e752c96f195111f/src/net/http/server.go#L1669-L1674). This means that `go-ipfs` will generally not see paths including `.` and `..` when referenced via a URI, whether `http:` or `fs:` or `file:`, rendering these paths equivalent for most users already. POSIX expects `.` and `..` to have specific meanings which can be satisfied using this logic. Rob Pike's [Lexical File Names in Plan 9 or Getting Dot-Dot Right](https://web.archive.org/web/20150205025854/http://plan9.bell-labs.com/sys/doc/lexnames.html) is also applicable to IPFS, describing detailed rationale for treating Plan 9's non-hierarchical paths in this manner, stating in part:

> The ability to construct union directories and other intricate naming structures introduces some thorny problems: as with symbolic links, the name space is no longer hierarchical, files and directories can have multiple names, and the meaning of `..`, the parent directory, can be ambiguous.
>
> The meaning of `..` is straightforward if the directory is in a locally hierarchical part of the name space, but if we ask what `..` should identify when the current directory is a mount point or union directory or multiply symlinked spot (which we will henceforth call just a mount point, for brevity), there is no obvious answer. Name spaces have been part of Plan 9 from the beginning, but the definition of `..` has changed several times as we grappled with this issue.
> …
> Frustrated by this situation, and eager to have better-defined names for some of the applications described later in this paper, we recently proposed the following definition for `..`:
> The parent of a directory `X`, `X/..`, is the same directory that would obtain if we instead accessed the directory named by stripping away the last path name element of `X`.

Why should string `/ipfs/base58hash/foo/../bar` not be immediately resolved to Path `/ipfs/base58hash/bar`? Are we entertaining the idea that it could have any other interpretation?